Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/BrushTool.kt

```diff
@@ -0,0 +1,171 @@
+ */
+package org.catrobat.paintroid.tools.implementation
+
+import android.graphics.Canvas
+import android.graphics.Paint
+import android.graphics.Path
+import android.graphics.PointF
+import androidx.annotation.VisibleForTesting
+import org.catrobat.paintroid.command.CommandManager
+import org.catrobat.paintroid.tools.ContextCallback
+import org.catrobat.paintroid.tools.Tool.StateChange
+import org.catrobat.paintroid.tools.ToolPaint
+import org.catrobat.paintroid.tools.ToolType
+import org.catrobat.paintroid.tools.Workspace
+import org.catrobat.paintroid.tools.common.CommonBrushChangedListener
+import org.catrobat.paintroid.tools.common.CommonBrushPreviewListener
+import org.catrobat.paintroid.tools.common.MOVE_TOLERANCE
+import org.catrobat.paintroid.tools.options.BrushToolOptionsView
+import org.catrobat.paintroid.tools.options.ToolOptionsVisibilityController
+import kotlin.math.abs
+import kotlin.math.max
+
+open class BrushTool(
+    private val brushToolOptionsView: BrushToolOptionsView,
+    contextCallback: ContextCallback,
+    toolOptionsViewController: ToolOptionsVisibilityController,
+    toolPaint: ToolPaint,
+    workspace: Workspace,
+    commandManager: CommandManager
+) : BaseTool(contextCallback, toolOptionsViewController, toolPaint, workspace, commandManager) {
+    protected open val previewPaint: Paint
+        get() = toolPaint.previewPaint
+
+    protected open val bitmapPaint: Paint
+        get() = toolPaint.paint
+
+    override val toolType: ToolType
+        get() = ToolType.BRUSH
+
+    @VisibleForTesting
+    @JvmField
+    var pathToDraw: Path = Path()
+    private var initialEventCoordinate: PointF? = null
+    private var pathInsideBitmap = false
+    private val drawToolMovedDistance = PointF(0f, 0f)
+
+    init {
+        pathToDraw.incReserve(1)
+        brushToolOptionsView.setBrushChangedListener(CommonBrushChangedListener(this))
+        brushToolOptionsView.setBrushPreviewListener(
+            CommonBrushPreviewListener(
+                toolPaint,
+                toolType
+            )
+        )
+        brushToolOptionsView.setCurrentPaint(toolPaint.paint)
+    }
+
+    override fun draw(canvas: Canvas) {
+        canvas.run {
+            save()
+            clipRect(0, 0, workspace.width, workspace.height)
+            drawPath(pathToDraw, previewPaint)
+            restore()
+        }
+    }
+
+    override fun handleDown(coordinate: PointF?): Boolean {
+        coordinate ?: return false
+        initialEventCoordinate = PointF(coordinate.x, coordinate.y)
+        previousEventCoordinate = PointF(coordinate.x, coordinate.y)
+        pathToDraw.moveTo(coordinate.x, coordinate.y)
+        drawToolMovedDistance.set(0f, 0f)
+        pathInsideBitmap = workspace.contains(coordinate)
+        return true
+    }
+
+    override fun handleMove(coordinate: PointF?): Boolean {
+        if (initialEventCoordinate == null || previousEventCoordinate == null || coordinate == null) {
```

Comment:
Move the condition to a function, please. (Duplicated in handleUp)

Reply:
sure

Reply:
done :)